src/mono/mono/mini/mini.c

```diff
@@ -4591,6 +4591,9 @@ mini_get_simd_type_info (MonoClass *klass, guint32 *nelems)
 	} else if (!strcmp (klass_name, "Vector2")) {
 		*nelems = 2;
 		return MONO_TYPE_R4;
+	} else if (!strcmp (klass_name, "Vector3")) {
+		*nelems = 4;
```

Comment:
Shouldn't this be `*nelems = 3;` ?
On that note, see my comment on `mini_number_of_elements`

Reply:
That's the reason it was split into two methods. We want Vector 3 to be Vector4 for all calculations (3 elements + zero). LLVM uses this function to determine the type of the vector so if we return 3 here it will create vector 3x float instead of 4 x float. 

The only place might need Vector3 to be actually 3 elements is vector sum calculation for ARM64 (non -llvm). As it would return wrong `element_size`
https://github.com/dotnet/runtime/blob/b76ef7fa73c758daaa663a4d36d78d536df152fe/src/mono/mono/mini/simd-intrinsics.c#L653-L657

I do agree two functions might be a bit too many, and modifying the code of vector sum calculation might be more reasonable

Reply:
Thank you for the explanation (and resolving the `mini_number_of_elements`)